generator.go

```diff
@@ -220,28 +220,46 @@ func (g *Gen) NewV5(ns UUID, name string) UUID {
 func (g *Gen) getClockSequence() (uint64, uint16, error) {
 	var err error
 	g.clockSequenceOnce.Do(func() {
-		buf := make([]byte, 2)
+		buf := make([]byte, 4)
 		if _, err = io.ReadFull(g.rand, buf); err != nil {
 			return
 		}
-		g.clockSequence = binary.BigEndian.Uint16(buf)
+		g.clockSequence = binary.BigEndian.Uint32(buf)
 	})
 	if err != nil {
 		return 0, 0, err
 	}
 
-	g.storageMutex.Lock()
-	defer g.storageMutex.Unlock()
-
+retry:
 	timeNow := g.getEpoch()
-	// Clock didn't change since last UUID generation.
-	// Should increase clock sequence.
-	if timeNow <= g.lastTime {
-		g.clockSequence++
+	timeHi := atomic.LoadUint64(&g.lastTime)
```

Comment:
It should be. sync.Once structs are 96 bits each, io.Reader as an interface is 2*word size, epochFunc and hwAddrFunc are word-sized func pointers, which puts lastTimestamp at bit 320. Am I missing something?

Regardless of the above I agree that the PR should have this at the very least mentioned (in the struct declaration?) if it were to get merged. I'll wait for additonal feedback before I commit further.

Reply:
@alcore would it be fair to say one risk would be if the standard library changed the `sync.Once` struct? Would we  be able to mitigate that risk by moving the `lastTime` field to the beginning of the struct definition?

Reply:
The structure of `sync.Once` *should* be irrelevant to the alignment unless it suddenly starts containing a 128-bit value. But this concern then applies to just about any struct member. *If* Go ever ran on a platform with 128-bit memory addressing, the function pointers would become 128-bit sized and the compiler would be using 128-bit alignment for the struct - assuming current rules.

I think my previous comment might've been a bit misleading here. Alignment is inferred from the largest sized field of a struct (as per https://golang.org/ref/spec#Size_and_alignment_guarantees). In our case that's precisely the timestamp. From the top of my head - if it were stored as a pointer directly (instead of being referenced for the CAS in NewV1), on 32-bit platforms it would be a 32-bit value and therefore drop the entire struct's alignment to 32-bits since no other field would then be larger than a 32-bit word. 

My previous statement made the mistake of counting the memory offset and potential padding (i.e. there is none) instead of alignment, as - ironically - introducing the uint64 actually aligns the struct to 64-bits where previously it was 32-bit (on 32-bit), which could've increased its size due to padding.

*tl;dr*: `lastTime` can be moved to the beginning without any adverse effect and fall into the `The first word in (...) an allocated struct (...) can be relied upon to be 64-bit aligned.` guarantee, but as I understand https://golang.org/ref/spec#Size_and_alignment_guarantees - it is not necessary. It could even be moved to the very end - that would increase the size of the generator (padding after `hardwareAddr`), but as per Go's spec, the alignment would still be that of the largest member. I believe the comment in the `sync/atomic` docs is merely a reminder for people who juggle around with unsafe, raw memory pointers, since the 64-bit atomics are implemented as 32-bit words on 32-bit ARM IIRC.

Reply:
>  but as per Go's spec, the alignment would still be that of the largest member

This is not correct. The alignment of the beginning of the struct is equal to the alignment of the field with the most alignment (which is often not the largest field). There are no types on 32-bit platforms with 64-bit alignment. This creates a requirement that the sum of the sizes of the fields before it is a multiple of 8 bytes **and** that this struct is in its own allocation.

So we really do need to move `lastTime` to the beginning of the struct.

Reply:
An issue with my phrasing, where the suggestion that it can be moved to the end was segueing on a count of the preceding bits - it'd end up on bit 384 on 32-bit platforms (hardwareAddr being padded), which puts it on a 64-bit boundary again.

But yes, the fact this conversation is even taking place is proof enough that it should be moved to the 0-offset, along with the quote from the `sync/atomic` docs as a reminder why it's there :-) Makes it oh so much simpler.

Sidenote: Storing it as a pointer would also alleviate the requirements altogether, wouldn't it? It will get its own alloc once it escapes to the heap during struct construction, and by extension Go will allocate the value on a 64-bit boundary, while the pointer in the struct will take the alignment of the struct. I generally do just that, but didn't want to introduce a pointer to chase in this case.